test/functional/Valhalla/src/ValueObjectAttributeTestGenerator.java

```diff
@@ -0,0 +1,160 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2023
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+package org.openj9.test.lworld;
+
+import org.objectweb.asm.*;
+
+import static org.objectweb.asm.Opcodes.*;
+
+public class ValueObjectAttributeTestGenerator extends ClassLoader {
+	private static ValueObjectAttributeTestGenerator generator = new ValueObjectAttributeTestGenerator();
+
+    public static Class<?> generateClassWithTwoPreloadAttributes(String name, String[] classList1, String[] classList2) throws Throwable {
+		byte[] bytes = generateClass(name, new Attribute[] {
+            new PreloadAttribute(classList1),
+            new PreloadAttribute(classList2)});
+        return generator.defineClass(name, bytes, 0, bytes.length);
+	}
+
+	public static Class<?> generateClassWithPreloadAttribute(String name, String[] classList) throws Throwable {
+		byte[] bytes = generateClass(name, new Attribute[] {new PreloadAttribute(classList)});
+		return generator.defineClass(name, bytes, 0, bytes.length);
+    }
+
+    public static Class<?> findLoadedTestClass(String name) {
+		return generator.findLoadedClass(name);
+	}
+
+    public static byte[] generateClass(String name, Attribute[] attributes) {
+		/**
+		 * Currently value type is built on JDK22, so use java file major version 66 for now.
+		 * If moved this needs to be incremented to the next class file version. The check in j9bcutil_readClassFileBytes()
+		 * against BCT_JavaMajorVersionShifted needs to be updated as well.
+		 */
+		int classFileVersion = 66;
+
+		ClassWriter classWriter = new ClassWriter(0);
+		classWriter.visit(classFileVersion, ACC_PUBLIC | ACC_FINAL, name, null, "java/lang/Object", null);
+
+		/* Generate base constructor which just calls super.<init>() */
+		MethodVisitor methodVisitor = classWriter.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
```

Comment:
The value type constructor is changed to `<vnew>`, see https://github.com/eclipse-openj9/openj9/issues/14959 and https://cr.openjdk.org/~dlsmith/jep401/jep401-20230519/specs/value-objects-jvms.html#jvms-2.9.4. 

Note that `<vnew>` is called invoked via invokestatic and it has a non-void return type.

Reply:
Wouldn't the classes being listed in the preload attribute be the value classes?

Reply:
The file is ValueObjectAttributeTestGenerator.java, so I was assuming the generated class is a value class. After a second look, I don't see `ACC_VALUE_TYPE` used to generate the class. So `<init>` makes sense in this case. Maybe we can add another case that the generated class is a value class. 

> Wouldn't the classes being listed in the preload attribute be the value classes?

The spec does not force the class in the preload list to be value class. But I believe value class is the most popular use case. In your code PreloadClass is a non-value class. Maybe you can add another case the preload class is a value class:

`value class ValuePreloadClass {}`

Reply:
You're right the name doesn't make sense, I'm changing it to "ValhallaAttributeGenerator" etc.

I have added in the new test.